File: Review-03/Example-8-Python/room.py
from dataclasses import dataclass


class Flooring:
    def __init__(self, type_n: str = "Generic", unit_c: float = 1.0):
        self.type_name = type_n
        self.unit_cost = unit_c

    @property
    def unit_cost(self) -> float:
        return self.__unit_cost

    @unit_cost.setter
    def unit_cost(self, uc: float):
        self.__unit_cost = float(uc)


@dataclass
class DimensionSet:
    length: float = 1.0
    width: float = 1.0


class Room:

    def __init__(self, nme: str = "Generic"):
        self.__name = nme
        self.__dimensions = DimensionSet()
        self.__flooring = Flooring()

    @property
    def name(self):
        """
        Get the name using a property.

        @property must come before @...setter...
        """

        return self.__name

    def with_name(self, nme: str):
        """
        Set the name using the builder pattern.

        Args:
            nme: room name
        """

        self.__name = nme

        return self

    def with_flooring(self, nme: str, unit_c: float):
        """
        Set the Flooring using the builder pattern.

        Args
            nme: flooring type name
            unit_c: unit cost

        """

        self.__flooring.type_name = nme
        self.__flooring.unit_cost = unit_c

        return self

    def with_dimensions(self, l: float, w: float):
        """
        Set the Flooring using the builder pattern.

        Args:
            l: length
            w: width

        """

        self.__dimensions.length = l
        self.__dimensions.width = w

        return self

    def set_flooring(self, nme: str, unit_c: float):
        """
        Set the flooring.

        Args:
            nme: flooring type name
            unit_c: unit cost

        """

        self.__flooring.type_name = nme
        self.__flooring.unit_cost = unit_c

    def area(self) -> float:
        """
        Compute the area of flooring for a room.
        """

        return float(self.__dimensions.width) * self.__dimensions.length

    def flooring_cost(self) -> float:
        """
        Compute the flooring cost based on `self.area()` and unit cost.
        """

        return self.area() * self.__flooring.unit_cost

    def __str__(self) -> str:

        return f"Room ({self.name})\n" \
             + "  {:<6}: {:>8.1f}\n".format("Length", float(self.__dimensions.length)) \
             + "  {:<6}: {:>8.1f}\n".format("Width", float(self.__dimensions.width)) \
             + "  {:<6}: {:>8.1f}\n".format("Area", self.area()) \
             + "\n" \
             + f"  Flooring  : {self.__flooring.type_name}\n" \
             + f"  Unit Cost : $ {self.__flooring.unit_cost:>8.2f}\n" \
             + "  Total Cost: $ {:>8.2f}\n".format(self.flooring_cost())

    def __lt__(self, rhs) -> bool:

        if self.name == rhs.name:
            return self.area() < rhs.area()  # pylint caught the missing return

        return self.name < rhs.name

    def __eq__(self, rhs) -> bool:
        if not isinstance(rhs, Room):
            return False

        return self.name == (rhs.name) and self.area() == rhs.area()

File: Review-99-Lecture-Snippets/Rust-Builder/src/room.rs
use std::cmp::Ordering;
use std::fmt;
use std::fmt::Display; //,Formatter,Result};
use crate::flooring::*;


#[derive(Clone)]
pub struct DimensionSet {
    pub length: f64,
    pub width: f64,
}

impl DimensionSet {
    pub fn new(l: f64, w: f64) -> Self {
        DimensionSet {
            length: l,
            width: w,
        }
    }
}

impl Default for DimensionSet {
    fn default() -> Self {
        DimensionSet::new(1f64, 1f64)
    }
}

//------------------------------------------------------------------------------
#[derive(Clone)]
pub struct Room {
    pub name: String,
    pub dimensions: DimensionSet,
    pub flooring: Flooring,
}

impl Room {
    /// Set the name using the builder pattern.
    ///
    /// # Arguments
    ///  * `nme` - room name
    ///
    pub fn with_name(mut self, nme: &str) -> Self {
        self.name = nme.to_string();

        self
    }

    /// Set the Flooring using the builder pattern.
    ///
    /// # Arguments
    ///  * `nme` - flooring type name
    ///  * `unit_c` - unit cost
    ///
    pub fn with_flooring(mut self, nme: &str, unit_c: f64) -> Self {
        self.flooring.type_name = nme.to_string();
        self.flooring.unit_cost = unit_c;

        self
    }

    /// Set the Flooring using the builder pattern.
    ///
    /// # Arguments
    ///  * `l` - length
    ///  * `w` - width
    ///
    pub fn with_dimensions(mut self, l: f64, w: f64) -> Self {
        self.dimensions.length = l;
        self.dimensions.width = w;

        self
    }

    /// Set the flooring.
    ///
    /// # Arguments
    ///  * `nme` - flooring type name
    ///  * `unit_c` - unit cost
    ///
    pub fn set_flooring(&mut self, nme: &str, unit_c: f64) {
        self.flooring.type_name = nme.to_string();
        self.flooring.unit_cost = unit_c;
    }

    /// Compute the area of flooring for a room.
    pub fn area(&self) -> f64 {
        self.dimensions.width * self.dimensions.length
    }

    /// Compute the flooring cost based on `self.area()` and unit cost.
    pub fn flooring_cost(&self) -> f64 {
        self.area() * self.flooring.unit_cost
    }
}

impl Default for Room {
    fn default() -> Self {
        Room {
            name: "Generic".to_string(),
            dimensions: Default::default(),
            flooring: Default::default(),
        }
    }
}

impl Display for Room {
    #[allow(unused_must_use)]
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        writeln!(f, "Room ({})", self.name);
        writeln!(f, "  {:<6}: {:>8.1}", "Length", self.dimensions.length);
        writeln!(f, "  {:<6}: {:>8.1}", "Width", self.dimensions.width);
        writeln!(f, "  {:<6}: {:>8.1}", "Area", self.area());
        writeln!(f);
        writeln!(f, "  Flooring  : {}", self.flooring.type_name);
        writeln!(f, "  Unit Cost : $ {:>8.2}", self.flooring.unit_cost);
        writeln!(f, "  Total Cost: $ {:>8.2}", self.flooring_cost())
    }
}

impl PartialOrd for Room {
    fn partial_cmp(&self, rhs: &Self) -> Option<Ordering> {
        if self.name.eq(&rhs.name) {
            self.area().partial_cmp(&rhs.area())
        }
        else {
            self.name.partial_cmp(&rhs.name)
        }
    }
}

impl PartialEq for Room {
    fn eq(&self, rhs: &Self) -> bool {
        return self.name.eq(&rhs.name)
            && self.area().eq(&rhs.area());
    }
}


//------------------------------------------------------------------------------
pub struct RoomBuilder<'a> {
    name: &'a str
}

impl<'a> RoomBuilder<'a> {
    pub fn with_name(mut self, nme: &str) -> Self {

        self
    }

    pub fn with_flooring(mut self, nme: &str, unit_c: f64) -> Self {

        self
    }

    pub fn with_dimensions(mut self, l: f64, w: f64) -> Self {

        self
    }

    pub fn build(mut self) -> Option<Room> {
        None
    }
}

impl From<(f64, f64)> for DimensionSet {
    fn from(dims: (f64, f64)) -> Self {
        DimensionSet::new(dims.0, dims.1)
    }
}
